drivers: spi: siwx91x: Add siwx91x SPI primary driver

smalae opened the pull request:
Implement SPI primary driver for siwx91x device

Martinhoff-maker commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,687 @@
+				   struct gspi_siwx91x_dma_channel *channel, uint32_t block_count,
+				   bool is_tx, uint8_t dfs)
+{
+	struct dma_config cfg = {
```
How did you manage to know that the DMA transfer will happen on the SPI request ? Normally, there is the dma_slot parameter for this. from what I understand, It is hard linked to the channel number ?

smalae replied:
Yes, in the SiWx917, peripherals are indeed hard-linked to specific DMA channel numbers. This means that each peripheral is pre-assigned to a fixed DMA channel.

Martinhoff-maker commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,687 @@
+
+		/* Configure RX DMA channel */
+		if (data->dma_rx.chan_nb == GSPI_RX_UDMA_CH_NUM) {
+			channel_filter = GSPI_RX_UDMA_CH_NUM;
```
from what I see in the SIWX91x dma driver, chan request with chan filter is not implemented ? are you sure you get the right channel if channel are hard linked to dma request ? That's why you recheck behind ?

smalae replied:
The channel filter functionality was recently implemented for the SiWx917 DMA driver. I am rechecking it to ensure that no other thread is already using the same channel. This additional validation step is crucial, as concurrent usage of the channel by another thread could lead to the channel being unallocated

jerome-pouiller commented:
Hello @smalae,
The commit is large enough and touch to several components. Could you split it in atomic parts:
  - changes in the the clock framework
  - the SPI driver by itself
  - the changes in the DT
  - tests additions

Martinhoff-maker commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,687 @@
+#define GSPI_DMA_MAX_DESCRIPTOR_TRANSFER_SIZE 1024
+#define GSPI_TX_UDMA_CH_NUM                   11
+#define GSPI_RX_UDMA_CH_NUM                   10
+#define GSPI_CLK_EN_BIT_FIELD                 0
```
Are some of these define are not available in a file from wiseconnect ? like gspi low level driver header ?

smalae replied:
Unfortunately, these definitions were not included in the WiseConnect SDK. Despite my efforts to locate and fetch them, I was unable to find them.

jerome-pouiller commented on dts/arm/silabs/siwg917.dtsi:
```diff
@@ -296,6 +296,17 @@
 			status = "disabled";
 		};
+
+		gspi: spi@45030000 {
```
I rather use the generic names for the DT labels: `spi0: spi@45030000 {`

smalae replied:
Updated

smalae committed:
drivers: spi: siwx91x: SPI clock initialization for siwx91x

Clock driver changes required for initializing the SPI clock
for the siwx91x driver

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

smalae committed:
drivers: spi: siwx91x: Add siwx91x SPI primary driver

Implement SPI driver for siwx91x device

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

smalae committed:
drivers: spi: siwx91x: DTS changes for siwx91x SPI driver

1. Create a YAML file for SPI node
2. Add SPI node in the siwx917.dtsi

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

smalae committed:
drivers: spi: siwx91x: spi_loopback test support

1. Add board config and overlay files
2. Update testcase.yaml

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

smalae commented on drivers/clock_control/clock_control_silabs_siwx91x.c:
```diff
@@ -77,6 +78,11 @@ static int siwx91x_clock_on(const struct device *dev, clock_control_subsys_t sys
 		break;
+	case SIWX91X_CLK_GSPI:
+		RSI_PS_M4ssPeriPowerUp(M4SS_PWRGATE_ULP_EFUSE_PERI);
+		RSI_CLK_PeripheralClkEnable3(M4CLK, M4_SOC_CLK_FOR_OTHER_ENABLE);
```
This section is responsible for enabling the GSPI peripheral clock. According to the GSPI initialization sequence in the rsi_driver, this step is part of the clock configuration process. The clock configuration involves three sets of registers, each corresponding to different peripherals, with GSPI belonging to the third set. For further details on the clock configuration process, you can refer to the implementation in rsi_pll.c

smalae replied:
Yes, I don't think it's required here. Removed RSI_CLK_PeripheralClkEnable3() call

smalae commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,687 @@
+#define GSPI_MAX_BAUDRATE_FOR_POS_EDGE_SAMPLE 40000000
+#define GSPI_DMA_MAX_DESCRIPTOR_TRANSFER_SIZE 1024
+#define GSPI_TX_UDMA_CH_NUM                   11
+#define GSPI_RX_UDMA_CH_NUM                   10
```
After passing data->dma_rx.dma_dev and data->dma_tx.dma_dev to the dma_request_channel() API, I am rechecking the values of the channels to ensure that the filtered channel is correctly allocated and available for the transfer.

Martinhoff-maker replied:
Understand but then the DTS defined channel should be the true. You may request the DTS channel and if it failed, you can't use GPSI because you don't have the access to the right channel.

smalae replied:
Understood, validating the channel number from DTS instead.

smalae commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,687 @@
+
+	spi_context_cs_control(ctx, true);
+
+	sys_write32(BIT(data->dma_rx.chan_nb), (mem_addr_t)&dma_reg_base->CHNL_ENABLE_SET);
```
That implementation introduces a slight delay between the start of the RX channel and the TX channel. This delay, although minimal, is sufficient to cause the loss of the first byte of data during transmission

Martinhoff-maker replied:
I don't get the point, if you start RX first, you shouldn't loss any data ? Moreover, using directly the DMA base register in SPI driver is not correct IMO.

smalae replied:
Thanks for pointing this out. After debugging further, I identified this as a bug in DMA, which was fixed recently. Now dma_start() is being used here.

smalae commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,687 @@
+#include <zephyr/logging/log.h>
+#include "clock_update.h"
+
+LOG_MODULE_REGISTER(spi_max32, CONFIG_SPI_LOG_LEVEL);
```
Corrected

smalae commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,687 @@
+#ifdef CONFIG_SPI_SILABS_SIWX91X_GSPI_DMA
+/* Structure for DMA configuration */
+struct gspi_siwx91x_dma_channel {
+	/* DMA device pointer */
```
Removed unwanted comments

smalae commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,687 @@
+		SPI_CONTEXT_INIT_SYNC(gspi_data_##inst, ctx),                                      \
+		SPI_CONTEXT_CS_GPIOS_INITIALIZE(DT_DRV_INST(inst), ctx)                            \
+			SPI_SILABS_SIWX91X_GSPI_DMA_CHANNEL(inst, rx)                              \
+				SPI_SILABS_SIWX91X_GSPI_DMA_CHANNEL(inst, tx)};                    \
```
Updated

jerome-pouiller commented on drivers/spi/Kconfig.silabs_siwx91x_gspi:
```diff
@@ -0,0 +1,28 @@
+if SPI_SILABS_SIWX91X_GSPI
+
+config SPI_SILABS_SIWX91X_GSPI_DMA
+	bool "SIWX91X MCU GSPI DMA Support"
```
Since this is a sub-option of `SPI_SILABS_SIWX91X_GSPI`, we can just say `DMA Support` (that's said, we have to clean up the names of all our drivers)

smalae replied:
Updated

jerome-pouiller commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,690 @@
+#ifdef CONFIG_SPI_SILABS_SIWX91X_GSPI_DMA
+	struct gspi_siwx91x_dma_channel dma_rx;
+	struct gspi_siwx91x_dma_channel dma_tx;
+#endif /* CONFIG_SPI_SILABS_SIWX91X_GSPI_DMA */
```
The comment is is not required if the condition is obvious.

smalae replied:
Removed

jerome-pouiller commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,690 @@
+	int chan_nb;
+	struct dma_block_config dma_descriptors[CONFIG_SPI_SILABS_SIWX91X_GSPI_DMA_MAX_BLOCKS];
+};
+#endif /* CONFIG_SPI_SILABS_SIWX91X_GSPI_DMA */
```
You can define this struct unconditionally.

smalae replied:
Done

jerome-pouiller commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,690 @@
+	int ret;
+#ifdef CONFIG_SPI_SILABS_SIWX91X_GSPI_DMA
+	int channel_filter;
+#endif /* CONFIG_SPI_SILABS_SIWX91X_GSPI_DMA */
```
You can use `__maybe_unused` rather than this `#ifdef`.

smalae replied:
Done

jerome-pouiller commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,690 @@
+#define USE_PREV_LENGTH_BIT_FIELD             7
+#define GSPI_BUSY_BIT_FIELD                   0
+#define SPI_HIGH_PERFORMANCE_EN_BIT_FIELD     11
+#define GSPI_MANUAL_CSN_BIT_FIELD             0
```
I assume your idea was to use:
```
    sys_set_bit((mem_addr_t)&cfg->reg->GSPI_CLK_CONFIG, GSPI_CLK_SYNC_BIT_FIELD);
```
rather than:
```
    cfg->reg->GSPI_CLK_CONFIG_b.GSPI_CLK_SYNC = 1;
```

Unfortunately, I believe this require to redefine too many defines. I believe we should rather use the second option.

smalae replied:
Okay, updated accordingly.

jerome-pouiller commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,690 @@
+
+	if (0 != (spi_cfg->operation & (SPI_MODE_CPOL | SPI_MODE_CPHA)) &&
+	    ((SPI_MODE_CPOL | SPI_MODE_CPHA) !=
+	     (spi_cfg->operation & (SPI_MODE_CPOL | SPI_MODE_CPHA)))) {
```
hmm... how to improve this condition? Maybe
```
        if (!!(spi_cfg->operation & SPI_MODE_CPOL) != !!(spi_cfg->operation & SPI_MODE_CPHA)) {
```
or
```
        if ((spi_cfg->operation & SPI_MODE_CPOL) ^^ (spi_cfg->operation & SPI_MODE_CPHA)) {
```
(matter of taste)

smalae replied:
I prefer the first one. Updated.

jerome-pouiller commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,690 @@
+	sys_write32(clk_div_factor, (mem_addr_t)&cfg->reg->GSPI_CLK_DIV);
+
+	/* Configure SPI clock mode */
+	if (0 == (spi_cfg->operation & (SPI_MODE_CPOL | SPI_MODE_CPHA))) {
```
Avoid [Yoda notation][1]

[1]: https://en.wikipedia.org/wiki/Yoda_conditions#Criticism

smalae replied:
Updated

jerome-pouiller commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,690 @@
+		dma_release_channel(data->dma_tx.dma_dev, data->dma_tx.chan_nb);
+
+		/* Configure RX DMA channel */
+		if (data->dma_rx.chan_nb == GSPI_RX_UDMA_CH_NUM) {
```
This only works on `siwx91x`. If the GSPI is used in another soc, this value will probably change. This is the job of the DT to provide this value. The driver can't verify it.

smalae replied:
Fixed

jerome-pouiller commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,690 @@
+		}
+
+		/* Configure TX DMA channel */
+		if (data->dma_tx.chan_nb == GSPI_TX_UDMA_CH_NUM) {
```
See my previous comment.

smalae replied:
Fixed

jerome-pouiller commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,690 @@
+		}
+
+		transaction_len -=
+			gspi_siwx91x_fill_desc(cfg, desc, NULL, transaction_len, is_tx, dfs);
```
Nitpick: we prefer to break between the arguments (as you did with `gspi_siwx91x_fill_desc()`):
```
		transaction_len -= gspi_siwx91x_fill_desc(cfg, desc, NULL,
							  transaction_len, is_tx, dfs);
```

smalae replied:
Updated

jerome-pouiller commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,690 @@
+
+	if (spi_context_tx_buf_on(&data->ctx)) {
+		if (dfs == 1) {
+			tx_frame = UNALIGNED_GET((uint8_t *)(data->ctx.tx_buf));
```
`UNALIGNED_GET()` is not required with `uint8_t`.

smalae replied:
Removed for uint8_t

jerome-pouiller commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,690 @@
+	uint16_t rx_frame;
+
+	tx_frame = gspi_siwx91x_next_tx(data, dfs);
+	gspi_siwx91x_send(cfg, (uint32_t)tx_frame);
```
I think the cast is not required.

smalae replied:
Updated

tbursztyka commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,690 @@
+		ret = gspi_siwx91x_transceive_dma(dev, config);
+	} else if (!asynchronous) {
+		/* Perform synchronous polling transceive */
+		ret = gspi_siwx91x_transceive_polling_sync(dev, &data->ctx);
```
spi_transceive_async() is not about being irq based (or in reverse: spi_transceive() is not meant to be in polling mode).

The only difference on the 2 is that the _async will return asap, and use a callback to signal when the transfer is done, while the other is just a blocking call: but it does not mention anywhere that it should be polling. This is a common misunderstanding, I guess we'll need no detail that in the api file.

So remove the polling stuff, just use your dma version (an irq based if dma is not available would be nice).

smalae replied:
Thank you for the suggestion! I agree that using the DMA version for the sync API is the right approach. I've implemented this change, replacing the default polling mechanism for synchronous operations. However, switching to synchronous with IRQs, instead of DMA, presents a challenge. Large data transfers can trigger a lot of IRQs, which is not an issue when DMA is used. Is it acceptable to retain polling as a fallback for cases where the DMA component isn't available? This will allow users to quickly perform basic driver checks without requiring the DMA component.

jerome-pouiller commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,662 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT silabs_siwx91x_gspi
```
I haven't noticed that during my first review, but usually, the name of the driver is the vendor  (`silabs`) and followed by the hardware block name (ie. `silabs,ldma`). Generally, the device included on the `siwx91x` don't have any names. So we use generic name like `siwx91x-pwm`. (BTW, we used `siwx91x-dma` while the block is named `udma`)

If we follow these rule, the driver should be named either `silabs,gspi` or `silabs,siwx91x-spi`.

smalae replied:
Using silabs,gspi

jerome-pouiller commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,662 @@
+#include "clock_update.h"
+
+LOG_MODULE_REGISTER(spi_siwx91x_gspi, CONFIG_SPI_LOG_LEVEL);
+#include "spi_context.h"
```
I believe you can group this include with the others.

smalae replied:
spi_context.h has a dependency on log. Facing a build issue if included with others

jerome-pouiller commented on drivers/spi/Kconfig.silabs_siwx91x_gspi:
```diff
@@ -0,0 +1,28 @@
+	help
+	  One block is needed for every chunk found in the SPI transaction and every 1024 bytes
+
+endif #SPI_SILABS_SIWX91X_GSPI
```
Space after the `#`

smalae replied:
updated

jerome-pouiller commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,662 @@
+struct gspi_siwx91x_dma_channel {
+	const struct device *dma_dev;
+	int chan_nb;
+#ifdef CONFIG_SPI_SILABS_SIWX91X_GSPI_DMA
```
I think it make more sense to declare `dma_rx` and `dma_tx` in `#ifdef` rather than `dma_descriptors` (unless it break the build, but I believe it's fine).

smalae replied:
CONFIG_SPI_SILABS_SIWX91X_GSPI_DMA_MAX_BLOCKS comes only when CONFIG_SPI_SILABS_SIWX91X_GSPI_DMA is enabled. Facing a build issue if moved out #ifdef.

jerome-pouiller commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,662 @@
+static int gspi_siwx91x_config(const struct device *dev, const struct spi_config *spi_cfg,
+			       spi_callback_t cb, void *userdata)
+{
+	[[maybe_unused]] struct gspi_siwx91x_data *data = dev->data;
```
No one else use `[[maybe_unused]]`. `__maybe_unused` seems the common to declare that.

smalae replied:
updated

jerome-pouiller commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,662 @@
+
+	/* Configure clock divider based on the requested bit rate */
+	if (bit_rate > GSPI_MAX_BAUDRATE_FOR_DYNAMIC_CLOCK) {
+		clk_div_factor = GSPI_STATIC_CLOCK_DIV_FACTOR;
```
IMHO, `clk_div_factor = 1;` is more meaningful.

smalae replied:
done

jerome-pouiller commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,662 @@
+	}
+
+	if (clk_div_factor < 1) {
+		cfg->reg->GSPI_CLK_CONFIG_b.GSPI_CLK_EN = ENABLE;
```
`ENABLE`? I believe everyone understand that `1` means enable. (While when I read `ENABLE`, I have to check there is no tricks in the definition)

smalae replied:
done

jerome-pouiller commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,662 @@
+	}
+
+	/*  Update the number of Data Bits */
+	sys_write32(SPI_WORD_SIZE_GET(spi_cfg->operation), (mem_addr_t)&cfg->reg->GSPI_WRITE_DATA2);
```
Since we already directly access to `cfg->reg->GSPI_CONFIG1_b`, we can do the same for `cfg->reg->GSPI_WRITE_DATA2`:

    cfg->reg->GSPI_WRITE_DATA2 = SPI_WORD_SIZE_GET(spi_cfg->operation);

(yeah, I have relaxed my opinion on the usage of `sys_write32()`)

smalae replied:
Updated

jerome-pouiller commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,662 @@
+	if (!spi_context_configured(&data->ctx, config)) {
+		ret = gspi_siwx91x_config(dev, config, cb, userdata);
+		if (ret) {
+			return ret;
```
missing `spi_context_unlock()`?

smalae replied:
Added

jerome-pouiller reviewed (commented):
The code is pleasant to read. My comments are mostly nitpicking.

smalae commented on drivers/spi/Kconfig.silabs_siwx91x_gspi:
```diff
@@ -0,0 +1,28 @@
+	  Enable DMA support for SIWX91X MCU GSPI driver.
+
+config SPI_SILABS_SIWX91X_GSPI_DMA_MAX_BLOCKS
+	int "Silabs Siwx91x GSPI controller driver maximum DMA transfer block per channel for a transaction."
```
Updated

smalae commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,662 @@
+		return ret;
+	}
+
+	ret = spi_context_wait_for_completion(&data->ctx);
```
Done

jerome-pouiller commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -0,0 +1,662 @@
+	/* Check if DMA is enabled */
+	if (spi_siwx91x_is_dma_enabled_instance(dev)) {
+		/* Perform DMA transceive */
+		ret = gspi_siwx91x_transceive_dma(dev, config);
```
I thought I had removed this comment. I believe the code is correct because `spi_context_wait_for_completion()` checks the `asynchronous` field (and that's reason why I asked for a comment above this call).
